src/tox_travis/toxenv.py

```diff
@@ -4,50 +4,50 @@
 import os
 import sys
 import re
-import py
-import tox.config
 
-from tox.config import _split_env as split_env
+from tox.config import _split_env as split_env, SectionReader
 try:
     from tox.config import default_factors
 except ImportError:
     default_factors = None
 
-from .utils import TRAVIS_FACTORS, parse_dict
+from .utils import TRAVIS_FACTORS, parse_dict, get_iniconfig
 
 
-def default_toxenv():
+def coerce_pypy_version():
+    # Travis virtualenv do not provide `pypy3`, which tox tries to execute.
+    # This doesnt affect Travis python version `pypy3`, as the pyenv pypy3
+    # is in the PATH.
+    # https://github.com/travis-ci/travis-ci/issues/6304
+    # Force use of the virtualenv `python`.
+    version = os.environ.get('TRAVIS_PYTHON_VERSION')
+    if version and default_factors and version.startswith('pypy3.3-5.2-'):
+        default_factors['pypy3'] = 'python'
+
+
+def default_toxenv(config):
     """Default TOXENV automatically based on the Travis environment."""
     if 'TRAVIS' not in os.environ:
         return
 
     if 'TOXENV' in os.environ:
         return  # Skip any processing if already set
 
-    config = py.iniconfig.IniConfig('tox.ini')
+    tox_ini = get_iniconfig(config)
 
     # Find the envs that tox knows about
-    declared_envs = get_declared_envs(config)
+    declared_envs = get_declared_envs(tox_ini)
 
     # Find all the envs for all the desired factors given
-    desired_factors = get_desired_factors(config)
+    desired_factors = get_desired_factors(tox_ini)
 
     # Reduce desired factors
     desired_envs = ['-'.join(env) for env in product(*desired_factors)]
 
     # Find matching envs
     matched = match_envs(declared_envs, desired_envs,
                          passthru=len(desired_factors) == 1)
-    os.environ.setdefault('TOXENV', ','.join(matched))
-
-    # Travis virtualenv do not provide `pypy3`, which tox tries to execute.
-    # This doesnt affect Travis python version `pypy3`, as the pyenv pypy3
-    # is in the PATH.
-    # https://github.com/travis-ci/travis-ci/issues/6304
-    # Force use of the virtualenv `python`.
-    version = os.environ.get('TRAVIS_PYTHON_VERSION')
-    if version and default_factors and version.startswith('pypy3.3-5.2-'):
-        default_factors['pypy3'] = 'python'
+    config.envlist = matched
```

Comment:
Will this do what we want? We need to make sure that if `TOXENV` is set, or `tox -e` is used, that those take precedence over our `default_toxenv`.